File: 프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    int curr_time = 0;
    stack<int> st;
    
    
    for (int i = 0; i < progresses.size(); i++){
        // 현재 작업 걸리는 시간 계산
        int time = (100 - progresses[i]) / speeds[i];
        if ((100 - progresses[i]) % speeds[i] > 0){
            time ++;
        }
        
        // 현재 시간보다 더 걸림 -> stack에 있는 것 모두 뽑기
        if (curr_time < time){
            int count = 0;
            while (!st.empty() && st.top() <= curr_time){
                st.pop();
                count ++;
            }
            if (count > 0) answer.push_back(count);
            curr_time = time;
        }
        st.push(time);
        
    }
    answer.push_back(st.size());
    
    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        /**
        progress / speed = time
        */
        List<Integer> answer = new ArrayList<>();
        
        int size = progresses.length;
        int time = 0;
        int count = 0;
        
        for (int i = 0; i < size; i++) {
            int curr = (100 - progresses[i]) / speeds[i];
            if ((100 - progresses[i]) % speeds[i] > 0) {
                curr += 1;
            } 
            
            if (count > 0 && curr > time) {
                answer.add(count);
                count = 0;
            } 
            
            time = Math.max(time, curr);
            count += 1;
        }
        
        answer.add(count);
        
        return answer.stream()
            .mapToInt(Integer::intValue)
            .toArray();
    }
}